### ISSUE TITLE
pcs_constraint_order id collision

### ISSUE THREAD

**Issue by lefan** [OndrejHome/ansible.pcs-modules-2#13]
If we have playbook with such tasks then we have `id` collision (same id and one task update result of another).
```
- name: Start oreder db-x112-master and db-x112-master-vip
  pcs_constraint_order: resource1='db-x112-master'
                         resource1_action='promote'
                         resource2='db-x112-master-vip'
                         resource2_action='start'
                         symmetrical='false'
                         kind='Mandatory'
  run_once: true

- name: Stop oreder db-x112-master and db-x112-master-vip
  pcs_constraint_order: resource1='db-x112-master'
                         resource1_action='demote'
                         resource2='db-x112-master-vip'
                         resource2_action='stop'
                         symmetrical='false'
                         kind='Mandatory'
  run_once: true
```

**Comment by OndrejHome** [comment]
Hi @lefan, interesting. Let me check this. For completeness which version of 'pcs' is used on the cluster where this is run? ('pcs --version') I will assume that you use ansible 2.4 with python 2.7 as seen in other Issue that you have opened yesterday.

**Comment by lefan** [comment]
```
[root@x112-1 ~]# pcs --version
0.9.162
[root@x112-1 ~]# cat /etc/redhat-release 
CentOS Linux release 7.5.1804 (Core)
```

**Comment by OndrejHome** [comment]
Hi @lefan, 

I can reproduce this also on my setup and I have few ideas on what to do with it.
Issue seems to be in way of matching the order constraints on line 142

https://github.com/OndrejHome/ansible.pcs-modules-2/blob/7d12f5e9758e43242b2e09c338923d61276db331/library/pcs_constraint_order.py#L142-L143

This one detects first constraint with same order as trying to introduce, but ignores the actions during detection. Once it decides that this is the right one it will determine that actions are different.

https://github.com/OndrejHome/ansible.pcs-modules-2/blob/7d12f5e9758e43242b2e09c338923d61276db331/library/pcs_constraint_order.py#L172-L174

In such case it will take the ID of the constraint (which may be wrong as we have stopped on first match with same order of resources) and it will delete it. Then it will create the constraint resource should and may fail as there can be another constraint that is really a duplicate to this.

So I will test out few things, but mostly I think I will rework the way how the matching of order constraint is done. Thank you for reporting this issue!

Once I have some testing version I will get back to you.

**Comment by OndrejHome** [comment]
For now in `develop` I have idea on keeping old behaviour and introducing new default that matches also on resource actions. I'm not yet decided if the old behaviour will stay and would need to be somehow documented or fixed or if it will get removed. You can test out if with new default your use case would work or if there are any additional problems.

**Comment by lefan** [comment]
Hi @OndrejHome 
I have tested develop branch. Now it working.
Thank you for great work!

**Comment by lefan** [comment]
What do you think about other pcs_constraint_* libs? 
Can they be exposed to this problem?

**Comment by OndrejHome** [comment]
Thank you for testing @lefan!

The `pcs_constraint_colocation` seems to have same logic for declaring the constraint match as the `pcs_constraint_order` and I can replicate the colidsion of id's when I used something like below.

    - pcs_constraint_colocation: resource1='db-x112-master'
                             resource1_role='Master'
                             resource2='db-x112-master-vip'
                             score=200
      run_once: true
    - pcs_constraint_colocation: resource1='db-x112-master'
                             resource1_role='Slave'
                             resource2='db-x112-master-vip'
                             score=300
      run_once: true

`pcs constraint_location` doesn't have a support for roles so it seems to be not affected by this. (pcs has ability to specify role via `pcs <resource> rule ... role')

I will look on how to also address the `pcs_constraint_colocation` as it seems to be similar to `pcs_constraint_order`. If you hit the 'id collision' also in some other use case that you want to use, feel free to just append here. Seems that this was not yet area that was explored more in depth here :)

**TODO:**
- [ ] `pcs_constraint_order` - decide if to keep original behaviour and fix collision in it
- [ ] `pcs_constraint_colocation` - decide if to keep original behaviour and fix collision in it, try to address the collision in same way as for `pcs_constraint_order`

**Comment by OndrejHome** [comment]
Hi @lefan, Checking further I think it makes more sense to have fix as permanent functionality and not allow switch to old behaviour. I have updated this in the `devel` branch. Additionally I have also reworked the `pcs_constraint_colocation` in few ways that should now also consider the 'roles' that are passed to it for matching the constraint.  So now both `pcs_constraint_order` and `pcs_constraint_colocation` support only limited number of attributes for update(change) - list below. All other attributes are used for matching the constraint.
- `pcs_constraint_order`
  - matching attributes: resource names in given order + resource actions in given order
  - update attributes: `kind`, `symmetrical`
- `pcs_constraint_colocation`
  - matching attributes: resource names in given order + resource roles in given order
  - update attributes: `score`

I will keep these changes for a while in `develop` before merging into `master`. In case you have ability to test these new version and let me know if this works well for your use-cases, it would be great :)
If not, I will do some more testing myself just in case before merging to `master`.
